shopfloor_reception/actions/search.py

```diff
@@ -0,0 +1,14 @@
+# Copyright 2023 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl)
+
+from odoo.addons.component.core import Component
+
+
+class SearchAction(Component):
+
+    _inherit = "shopfloor.search.action"
+
+    def _get_origin_from_barcode(self, barcode):
+        if self.work.menu.scenario_id.key == "reception":
+            carriers = self.env["delivery.carrier"].search([])
+            return carriers._get_origin_from_barcode(barcode)
```

Comment:
There should be a super call

```suggestion
        if self.work.menu.scenario_id.key == "reception":
            carriers = self.env["delivery.carrier"].search([])
            return carriers._get_origin_from_barcode(barcode)
        return super()._get_origin_from_barcode(barcode)
```

also you have to adapt shopfloor_base/services/service.py https://github.com/OCA/wms/blob/fd917e896659271d1dfa36493510a1b6b0791dd0/shopfloor_base/services/service.py#L39

To something similar like this, otherwise there will be no "self.work.menu" in actions search
```
    def _actions_for(self, usage, **kw):
        propagate_kwargs = kw.get("propagate_kwargs", []) 
        if "menu" not in propagate_kwargs:
            propagate_kwargs.append("menu")
            kw["propagate_kwargs"] = propagate_kwargs
        return get_actions_for(self, usage, **kw)
```

Reply:
It should be like this.
If there is nothing found from the carrier, return the super result

```suggestion
        if self.work.menu.scenario_id.key == "reception":
            carriers = self.env["delivery.carrier"].search([])
            res = carriers._get_origin_from_barcode(barcode)
            if res:
                return res
        return super()._get_origin_from_barcode(barcode)
```

Reply:
true, I'll fix that

Reply:
@mmequignon Do you fix it? Or @TDu ?

Reply:
What is the state here?

Reply:
Thx. @TDu 
We also need to include 
https://github.com/OCA/wms/pull/822
as test-requirement otherwise this will not work because work has no attribute menu

Reply:
> There should be a super call
> 
> also you have to adapt shopfloor_base/services/service.py
> 
> https://github.com/OCA/wms/blob/fd917e896659271d1dfa36493510a1b6b0791dd0/shopfloor_base/services/service.py#L39
> 
> To something similar like this, otherwise there will be no "self.work.menu" in actions search
> 
> ```
>     def _actions_for(self, usage, **kw):
>         propagate_kwargs = kw.get("propagate_kwargs", []) 
>         if "menu" not in propagate_kwargs:
>             propagate_kwargs.append("menu")
>             kw["propagate_kwargs"] = propagate_kwargs
>         return get_actions_for(self, usage, **kw)
> ```

I think the actions should be kept independent from the shopfloor service implementation.
Instead of relying on a shopfloor menu, probably better to add a context key.

Reply:
But we have to otherwise, this will not work. It is only working with tests because they are not acting like, if it is called from the app. 
The problem is here https://github.com/OCA/wms/blob/d246abf384c343c69fb2ded4473164f7a43a4039/shopfloor_reception/tests/common.py#L101. Here the service is initialized with the menu from the start and https://github.com/OCA/wms/blob/d246abf384c343c69fb2ded4473164f7a43a4039/shopfloor_base/tests/common.py#L73 passes the menu arg also to https://github.com/OCA/wms/blob/d246abf384c343c69fb2ded4473164f7a43a4039/shopfloor_base/tests/common.py#L63 which initializes the data with _actions_for with the menu as kwarg. Thats why it is working in the tests but not within the app. 
@simahawk @jbaudoux @TDu please check this PR, it ensures that the menu is set on the data property.
If you have a look at https://github.com/OCA/wms/blob/d246abf384c343c69fb2ded4473164f7a43a4039/shopfloor_base/services/service.py#L202 which is called within the dispatch of service you can see that the menu is set after the init.

Reply:
But you don't think it is feasible with a context key ?

The real question, I raised,  is do we want to have the functions define in the actions tightly coupled with the shopfloor implementation, which I don't ?

Reply:
I think a context will also not Work. Of course moving the search into the service we would Not have to Deal with this problem at all. 👍🏼

Reply:
I agree w/ @TDu it shouldn't stay there.
If there's a difference in tests we have to fix that too.
IMO that method is specific to services and should stay in the base service mixin.

Reply:
Basically in the long run we want to extract all helper of Shopfloor in dedicated Odoo modules (that could provide these features in the Odoo backend), and Shopfloor will use these ones instead to perform some actions (extract_and_action_done, etc). It's not easy, but that's what we target.

Reply:
Indeed the context does not work, probably because it is a component. But I updated the code with an optional parameter (see the last fixup commit)

Reply:
ok then we should just move the method origin_move_from_scan to service mixin and the overwrite in the reception scenario.

Reply:
@simahawk  and @TDu  i updated https://github.com/OCA/wms/pull/822
To not have a menu in the action components while testing

Reply:
@TDu #822 is now merged can you do a rebase? Then the tests should fail

Reply:
I just did a rebase.
Hopefully the test will still be ok, because of my last fixup from last week